packages/expo-updates/android/src/main/java/expo/modules/updates/procedures/CheckForUpdateProcedure.kt

```diff
@@ -28,155 +27,154 @@ class CheckForUpdateProcedure(
 ) : StateMachineProcedure() {
   override val loggerTimerLabel = "timer-check-for-update"
 
-  override fun run(procedureContext: ProcedureContext) {
+  override suspend fun run(procedureContext: ProcedureContext) {
     procedureContext.processStateEvent(UpdatesStateEvent.Check())
 
-    AsyncTask.execute {
-      val embeddedUpdate = EmbeddedManifestUtils.getEmbeddedUpdate(context, updatesConfiguration)?.updateEntity
-      val extraHeaders = FileDownloader.getExtraHeadersForRemoteUpdateRequest(
-        databaseHolder.database,
-        updatesConfiguration,
-        launchedUpdate,
-        embeddedUpdate
-      )
-      databaseHolder.releaseDatabase()
-      fileDownloader.downloadRemoteUpdate(
-        extraHeaders,
-        object : FileDownloader.RemoteUpdateDownloadCallback {
-          override fun onFailure(e: Exception) {
-            procedureContext.processStateEvent(UpdatesStateEvent.CheckError(e.localizedMessageWithCauseLocalizedMessage()))
-            callback(IUpdatesController.CheckForUpdateResult.ErrorResult(e))
-            procedureContext.onComplete()
-          }
+    val embeddedUpdate = EmbeddedManifestUtils.getEmbeddedUpdate(context, updatesConfiguration)?.updateEntity
+    val extraHeaders = FileDownloader.getExtraHeadersForRemoteUpdateRequest(
+      databaseHolder.database,
+      updatesConfiguration,
+      launchedUpdate,
+      embeddedUpdate
+    )
+    databaseHolder.releaseDatabase()
+    fileDownloader.downloadRemoteUpdate(
+      extraHeaders,
+      object : FileDownloader.RemoteUpdateDownloadCallback {
+        override fun onFailure(e: Exception) {
+          procedureContext.processStateEvent(UpdatesStateEvent.CheckError(e.localizedMessageWithCauseLocalizedMessage()))
```

Comment:
do you know what's the fundamental difference between AsyncTask and coroutine, so we need to move the `processStateEvent` before callbacks?

Reply:
Coroutines don't block the thread they run on like AsyncTasks do. Most of those callbacks return a result to JS. Made more sense to me to update the state before doing that but I can revert it

Reply:
thanks for sharing that. yep updating native state before emitting js event makes sense to you. if we decide to do so, maybe to align the behavior between android and ios.